call-home/src/bin/callhome/main.rs

```diff
@@ -256,5 +296,91 @@ async fn generate_report(
     report.storage_media = StorageMedia::new(valid_disks);
 
     report.nexus = Nexus::new(event_data);
+
+    let logs = logs.lock().unwrap();
+    for log in logs.iter() {
+        let log_string = format!(
+            "[{} {} {}:{}] {}",
+            log.timestamp, log.level, log.file, log.line, log.message
+        );
+        report.logs.push(log_string);
+    }
     report
 }
+
+// Define the LogsLayer
+struct LogsLayer {
+    logs: Arc<Mutex<Vec<LogEntry>>>,
+}
+
+impl LogsLayer {
+    fn new(logs: Arc<Mutex<Vec<LogEntry>>>) -> Self {
+        LogsLayer { logs }
+    }
+}
+
+impl<S> Layer<S> for LogsLayer
+where
+    S: Subscriber,
+{
+    fn on_event(&self, event: &Event, _ctx: Context<S>) {
+        let timestamp = chrono::Utc::now().to_rfc3339();
+        let level = event.metadata().level();
+        let file = event.metadata().file().unwrap_or("").to_string();
+        let line = event.metadata().line().unwrap_or(0);
+
+        let mut visitor = LogVisitor::new();
+        event.record(&mut visitor);
+
+        let log_entry = LogEntry {
+            timestamp,
+            level: level.to_string(),
+            file,
+            line,
+            message: visitor.log,
+        };
+
+        // Only capture logs of level ERROR
+        if level == &Level::ERROR {
+            let mut logs = self.logs.lock().unwrap();
+            rev_length_check(&mut logs, ERR_LOG_BUF_CAPACITY - 1);
+            logs.push(log_entry);
+        }
+    }
+}
+
+//
+struct LogVisitor {
+    log: String,
+}
+
+impl LogVisitor {
+    fn new() -> Self {
+        LogVisitor { log: String::new() }
+    }
+}
+
+impl tracing::field::Visit for LogVisitor {
+    fn record_debug(&mut self, field: &tracing::field::Field, value: &dyn std::fmt::Debug) {
+        self.log
+            .push_str(&format!("{}: {:?};", field.name(), value));
+    }
+}
+
+#[derive(Debug, serde::Serialize)]
+struct LogEntry {
+    timestamp: String,
+    level: String,
+    file: String,
+    line: u32,
+    message: String,
+}
+
+/// Ensures the length of a vector is no more than the input limit, and removes elements from
+/// the front if it goes over the limit.
+fn rev_length_check<T>(logs: &mut Vec<T>, limit: usize) {
```

Comment:
Why not just something like this?

```rust
fn log_with_limit<T>(logs: &mut VecDeque<T>, message: T, limit: usize) {
    if logs.len() > limit-1 {
        logs.pop_front();
    }
    logs.push_back(message);
}
```

Reply:
Sure, using VecDeque across the board. Using this to limit size instead:
```rust
fn log_with_limit<T>(logs: &mut VecDeque<T>, message: T, limit: usize) {
    while logs.len().ge(&limit) {
        logs.pop_front();
    }
    logs.push_back(message);
}
```